Input Control visualization

nreese opened the pull request:
Adds visualization type for creating [input panels](https://github.com/elastic/kibana/issues/6884) for dashboards.
* `Range slider` input. Slider min and max obtained via a min/max aggregation call when panel is first rendered.
* `Options list` input. Options are loaded via a terms aggregation call when panel is first rendered.

<img width="1017" alt="screen shot 2017-08-31 at 11 48 53 am" src="https://user-images.githubusercontent.com/373691/29937330-6cab2928-8e42-11e7-8b43-d80fc54f72f6.png">

Replaces experimental PRs https://github.com/elastic/kibana/pull/11873 and https://github.com/elastic/kibana/pull/12126

nreese commented:
jenkins, test this

nreese commented:
jenkins, test this

Stacey-Gammon commented on src/core_plugins/input_control_vis/public/components/editor/field_select.js:
```diff
@@ -0,0 +1,73 @@
+import React, { Component } from 'react';
+import Select from 'react-select';
+
+export class FieldSelect extends Component {
```
This component looks like it could be generalized for the uiFramework as well - e.g. FilterableSelect.

@snide - what's your take on components like this. Would you prefer they get moved into the uiFramework ala k6, or should we hold off for now due to incoming k7?

Stacey-Gammon commented on src/core_plugins/input_control_vis/public/components/editor/list_control_editor.js:
```diff
@@ -0,0 +1,75 @@
+import { IndexPatternSelect } from './index_pattern_select';
+import { FieldSelect } from './field_select';
+
+export class ListControlEditor extends Component {
```
Make stateless function.  I thought we had eslint checks for this recently put it place, but maybe it's only turned on in the uiFramework...  CJ would know but I don't want to tag him on paternity leave.

Few other files/components I see that should be in the functional stateless variety.

Stacey-Gammon commented:
One more thing - can you add jest tests for all the react components?

weltenwort commented on src/core_plugins/input_control_vis/public/components/editor/control_editor.js:
```diff
@@ -0,0 +1,111 @@
+    if (this.props.controlParams.label) {
+      controlTitle = `${this.props.controlParams.type}: ${this.props.controlParams.label}`;
+    } else if (this.props.controlParams.fieldName) {
+      controlTitle = `${this.props.controlParams.type}: ${this.props.controlParams.fieldName}`;
```
Yes, I would apply that to props as well. In addition to the argument of primitives being easier to compare for `shouldComponentUpdate` checks, passing separate props also documents the external interface of the component better an an opaque "params" object. In general I found that many clean-code principles for function arguments are also applicable to component props.

kjbekkelund replied:
I don't think props needs to be as strict as state. I can think of several use-cases for objects and arrays in `props`, e.g. let's say you have a `Clusters` component that renders clusters: It should be able to receive an array of cluster objects, and maybe even do something like:

```js
return <ul>
  {
    clusters.map(cluster => <li><Cluster cluster={ cluster } /></li>)
  }
</ul>
```

Also, for large objects having to list everything in the component props would be a little heavy, I think.

When using objects or arrays, I think we should prefer to be more specific in the proptypes though. E.g. for an object we should often use `PropTypes.shape` (unless purely passing through to another component, where `PropTypes.object` might be okey as the underlying component would do a `PropTypes.shape`).

In the end it feels to me like a "it feels like we should use an object here" vs "it feels like we should use primitives here", which would be more case-by-case rather than strictly defined. (Might be difficult to be super-consistent with that, though)

Does that make sense?

archanid commented on src/core_plugins/input_control_vis/public/components/editor/control_editor.js:
```diff
@@ -0,0 +1,111 @@
+              <input
+                className="kuiTextInput"
+                type="text"
+                value={this.props.controlParams.label}
```
Are you sure about that? From working in x-pack-kibana I know that from 6.1 onward there is an eslint rule enforcing no spaces inside of `{}`. 
https://github.com/elastic/kibana/blob/master/packages/eslint-config-kibana/.eslintrc.js#L89-L100

Stacey-Gammon replied:
Whoa you are right (https://github.com/elastic/x-pack-kibana/pull/2371 and https://github.com/elastic/kibana/pull/13323 did all the conversions), sorry for the misdirection @nreese and thanks for pointing that out @archanid.  Looks like we switched it over to match the AirBnb style guide. I guess the react style guide should be updated because all of the examples in there use the old style of spaces inside.

Stacey-Gammon replied:
I filed https://github.com/elastic/kibana/issues/13846 to keep track.

simianhacker commented on src/core_plugins/input_control_vis/public/components/editor/index_pattern_select.js:
```diff
@@ -0,0 +1,50 @@
+import React, { Component } from 'react';
+import Select from 'react-select';
+
+export class IndexPatternSelect extends Component {
```
TSVB will probably always be adhoc because there are a lot of users with dashboards that depend on the current behavior; I don't want to take that away from them.

nreese commented on src/core_plugins/input_control_vis/public/components/editor/control_editor.js:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    return (
+      <div className="sidebar-item">
```
Not sure this gets you anything. With the above requested changes, control_editor is very straightforward.

timroes commented on src/core_plugins/input_control_vis/public/components/editor/control_editor.js:
```diff
@@ -0,0 +1,104 @@
+            Label
+          </label>
+          <div className="kuiSideBarFormRow__control kuiFieldGroupSection--wide">
+            <input
```
Please label this `input` with the above `label` (see [accessibility style guide](https://github.com/elastic/kibana/blob/master/style_guides/accessibility_guide.md#label-every-form-element)) using `htmlFor` in React.

timroes commented on src/core_plugins/input_control_vis/public/components/editor/control_editor.js:
```diff
@@ -0,0 +1,104 @@
+            {this.props.controlTitle}
+          </span>
+          <div className="vis-editor-agg-header-controls kuiButtonGroup kuiButtonGroup--united">
+            <button
```
This `button` is missing an `aria-label`, that describes it (since it doesn't have any textual content).

timroes commented on src/core_plugins/input_control_vis/public/components/editor/control_editor.js:
```diff
@@ -0,0 +1,104 @@
+            >
+              <i aria-hidden="true" className="fa fa-chevron-down" />
+            </button>
+            <button
```
This `button` is missing an `aria-label`, that describes it (since it doesn't have any textual content).

timroes commented on src/core_plugins/input_control_vis/public/components/editor/field_select.js:
```diff
@@ -0,0 +1,73 @@
+          Field
+        </label>
+        <div className="kuiSideBarFormRow__control kuiFieldGroupSection--wide">
+          <Select
```
Please label this `Select` (you can set an `id` with `inputProps={{ id: 'yourId' }}`) with the above label.

timroes commented on src/core_plugins/input_control_vis/public/components/editor/index_pattern_select.js:
```diff
@@ -0,0 +1,50 @@
+          Index Pattern
+        </label>
+        <div className="kuiSideBarFormRow__control kuiFieldGroupSection--wide">
+          <Select.Async
```
Please label this `Select` (see other comment).

timroes commented on src/core_plugins/input_control_vis/public/components/editor/list_control_editor.js:
```diff
@@ -0,0 +1,69 @@
+          Enable Multiselect
+        </label>
+        <div className="kuiSideBarFormRow__control">
+          <input
```
Please label this `input` with the above `label` (see other comment).

timroes commented on src/core_plugins/input_control_vis/public/components/editor/list_control_editor.js:
```diff
@@ -0,0 +1,69 @@
+          Size
+        </label>
+        <div className="kuiSideBarFormRow__control kuiFieldGroupSection--wide">
+          <input
```
Please label this `input` with the above `label` (see other comment).

timroes commented on src/core_plugins/input_control_vis/public/components/editor/range_control_editor.js:
```diff
@@ -0,0 +1,59 @@
+            Step Size
+          </label>
+          <div className="kuiSideBarFormRow__control kuiFieldGroupSection--wide">
+            <input
```
Please label this `input` with the above `label` (see other comment).

timroes commented on src/core_plugins/input_control_vis/public/components/editor/vis_editor.js:
```diff
@@ -0,0 +1,188 @@
+
+        <div className="kuiSideBarFormRow">
+          <div className="kuiSideBarFormRow__control kuiFieldGroupSection--wide">
+            <select
```
Please provide a label (e.g. via `aria-label`) for this `select`.

timroes commented on src/core_plugins/input_control_vis/public/components/vis/list_control.js:
```diff
@@ -0,0 +1,46 @@
+      <FormRow
+        label={this.props.control.label}
+      >
+        <Select
```
Please `label` this `Select` (see other comment).

timroes commented on src/core_plugins/input_control_vis/public/components/vis/range_control.js:
```diff
@@ -0,0 +1,49 @@
+        label={this.props.control.label}
+      >
+        <div className="inputRangeContainer">
+          <InputRange
```
Please label this element, using [`aria-labelledby`](https://github.com/davidchin/react-input-range#arialabelledby-string) on the `InputRange` and set this to the `id` of the label in `FormRow` (I guess that needs some modifications.

kobelb commented on src/core_plugins/input_control_vis/public/vis_controller.js:
```diff
@@ -0,0 +1,113 @@
+  initControls() {
+    this.controls = [];
+
+    controlFactory(
```
I'm thinking that if we change the `controlFactory` to deal with a single `control`, then we could switch from using a `callback` below for each individual control and use a `Promise` which seems more natural.

This would also allow us to use a `Promise.all` for each of the controlFactory to use with the `render` method.

nreese replied:
I did it this way so that a single `_msearch` would be used to init all controls in a panel. The weirdness comes from the fact that `fetch.these` uses the defer promise pattern

kobelb commented on src/core_plugins/kibana/public/dashboard/styles/index.less:
```diff
@@ -112,6 +112,10 @@ dashboard-grid {
   }
+
+  .gs-w {
+    z-index: auto;
```
Is something else setting this to something besides `auto`? `auto`  should be the default `z-index` for elements.

nreese replied:
Gridster sets this to 1 for each grid. This caused problems with overflow of the list select and needed to be set to auto

nreese commented on src/core_plugins/input_control_vis/public/components/editor/index_pattern_select.js:
```diff
@@ -0,0 +1,51 @@
+    this.loadOptions = this.loadOptions.bind(this);
+  }
+
+  loadOptions(input, callback) {
```
I was thinking of changing this to limit the index-pattern search to 100 results. Then I would configure Select.Async to fire the search every time a character is added to the filter. This is how the dashboard add 'Visualzation' and 'Saved Search' function. What do you think?

nreese commented on src/core_plugins/input_control_vis/index.js:
```diff
@@ -0,0 +1,9 @@
+export default function (kibana) {
```
Great idea. I have split the editor into two tabs; `Controls` and `Options`

The idea behind disabling "Update kibana filters on each change" by default is that only submitting all of the filters at a single instant puts the least amount of stress on the cluster. The default is designed for the usecase where the Panel contains many controls. Users can fiddle with several filters and then apply them all at once - so the dashboard only has to update on time instead of each time a control is modified.

timroes commented:
Unfortunately I think we still need some refactoring with the `id`s. `Id`s in HTML need to be unique on the whole page, meaning you cannot assign a static `id` in a component (like the editor components) that might get rendered multiple times.

You can use the `htmlIdGenerator` service from the `ui_framework`, to generate `id`s. I just merged this onto `master` (and `6.x`) so you would need to rebase on `master` again.

I explained how to use this service, in the [accessibility styleguide](https://github.com/elastic/kibana/blob/master/style_guides/accessibility_guide.md#how-to-generate-ids).

nreese commented:
@Stacey-Gammon I have added all of the requested jest tests and should have addressed all of your comments if you want to take another look.

@timroes I have fixed all of the html elements with id collisions.

kjbekkelund commented on src/core_plugins/input_control_vis/public/components/editor/control_editor.js:
```diff
@@ -0,0 +1,111 @@
+    this.state = {
+      isEditorCollapsed: true
+    };
+  }
```
When just setting the initial state, the constructor isn't needed:

```js
export class ControlEditor extends Component {

  state = {
    isEditorCollapsed: true
  }
```

kjbekkelund commented on src/core_plugins/input_control_vis/public/components/editor/range_control_editor.js:
```diff
@@ -0,0 +1,74 @@
+import { FieldSelect } from './field_select';
+
+function filterField(field) {
+  return ['number'].includes(field.type);
```
Can be simplified to

```js
return field.type === 'number'
```

Stacey-Gammon commented:
When I tried to create my first input control, the default tab is `Options`, not `Controls`.  Is that intentional?  I think it's a little confusing because all I see is a check box then and not much else to show me how to start setting it up.
<img width="1282" alt="screen shot 2017-09-12 at 9 10 05 am" src="https://user-images.githubusercontent.com/16563603/30327549-7383c9d8-979a-11e7-9921-94bfb6fd5421.png">

How difficult would it be to adjust the indexPattern drop down from allowing any dynamic index pattern, like TSVB?  I like the control here better in that it auto completes with your current patterns, but it might be nice to allow free form, if desired - as long as there isn't a technical reason not to. Then that will match what TSVB does.
<img width="347" alt="screen shot 2017-09-12 at 9 11 49 am" src="https://user-images.githubusercontent.com/16563603/30327627-c04a5ae8-979a-11e7-9550-7df9de96dfc1.png">

Stacey-Gammon commented:
I think I notice a bug with the index pattern drop down.

- Enter in a character that doesn't match any index pattern, e.g. '*'
<img width="347" alt="screen shot 2017-09-12 at 9 11 49 am" src="https://user-images.githubusercontent.com/16563603/30327722-1cbd34b2-979b-11e7-9732-b2a8318910bc.png">
- Click outside of the control which will clear the contents:
<img width="383" alt="screen shot 2017-09-12 at 9 17 18 am" src="https://user-images.githubusercontent.com/16563603/30327747-3848c200-979b-11e7-80c8-41c1b0fdea64.png">
- Press the drop down icon to see the possible index patterns, and nothing shows up.
<img width="225" alt="screen shot 2017-09-12 at 9 14 38 am" src="https://user-images.githubusercontent.com/16563603/30327769-44341e34-979b-11e7-9fc4-4279e6f5b554.png">

If you go back and add in another character, then explicitly delete it (not just clicking away), it will clear the value, and the next time you click the drop down icon, you'll see the expected patterns:
<img width="334" alt="screen shot 2017-09-12 at 9 18 38 am" src="https://user-images.githubusercontent.com/16563603/30327818-64820034-979b-11e7-9bc1-6ad16b825991.png">

Stacey-Gammon commented:
From a design standpoint, I think there should be padding above these controls:
<img width="402" alt="screen shot 2017-09-12 at 9 20 20 am" src="https://user-images.githubusercontent.com/16563603/30327877-a8439882-979b-11e7-94e7-71b7b802455c.png">

Stacey-Gammon commented:
The difference between `Clear` and `Cancel` isn't very clear.  Some help text on hover might help, though I'm not sure the words match what they do well. Also - `Clear` seems to always be enabled, even if there are no filters to clear.  Not a huge deal but makes it more confusing.  `Cancel` clears the current selection, right?  Maybe we should just make them `Clear selection` and `Clear filters`?  Or do we even need a `Clear selection`/`Cancel` button at all?   Isn't that what the little `x` does on the right?

One more thing - there is a delay when pressing `Clear` before the filters are actually cleared.  I wonder if there was some sort of visual we could add to make it clear that it worked - like maybe disabling the button and replacing the text with `Clearing...`.  Not sure the filter API will let you know when it's finished removing that filter though.

cc @snide 

 Another interaction that seems like a bug to me:
![message field clicking not updating](https://user-images.githubusercontent.com/16563603/30328387-3149c740-979d-11e7-9ace-49dcdde7c558.gif)
 The first message field only shows the bottom line, and when I click on it it seems to bring it into view, doesn't actually add it.  Maybe this is intentional, but it threw me off when I first tried it.  I thought the clicks were failing to go through.

Stacey-Gammon commented:
Looks like the input controls can also extend the borders of the dashboard panels, depending on their contents.  I think overflow needs to be scroll.

<img width="900" alt="screen shot 2017-09-12 at 9 39 45 am" src="https://user-images.githubusercontent.com/16563603/30328808-7237f6a4-979e-11e7-9b81-814ce24bb430.png">

btw, I notice now that `Cancel` is nice when you have a lot of input controls in one panel, since it clears all of them.  So maybe we keep it but think of different text. `Clear selections`, `Clear entries`? hmmm.  If we do keep Cancel, I think the hover text will help a lot.

btw, I think we also need more padding inside the dashboard panel:
<img width="771" alt="screen shot 2017-09-12 at 9 45 05 am" src="https://user-images.githubusercontent.com/16563603/30329061-3099076e-979f-11e7-81eb-b28e70c835db.png">

Stacey-Gammon commented:
The way this works with reporting is okay, but just to highlight some points/questions:

- Should we be putting these controls in a report at all, given that they are for manipulating the UI? Or does it make sense to show what filters are applied on a report?  Since we don't show the filters now, I think that last point might be no, but we may have more input controls that do have a useful UI. I think how it works now (including them) is probably our best option, even though it's a bit awkward.

- You can update filters in view mode, which updates the control, and doesn't require you to save before clicking 'generate report'. This can lead to reports where the input control fields don't match what the user saw then they pressed `generate report`. I suppose not a big deal since these are somewhat useless for reports anyway, we can live with this bug/oddity, given the benefit of the control.

- If the control has enough stuff in it, it might get cut off in the report.  I think we can also live with this, given this is the same issue we have with saved searches, and again, no one will probably care about input controls in a report anyway.

Stacey-Gammon commented on src/core_plugins/input_control_vis/public/components/editor/controls_tab.js:
```diff
@@ -0,0 +1,176 @@
+            <ListControlEditor
+              controlIndex={controlIndex}
+              controlParams={controlParams}
+              handleIndexPatternChange={this.handleIndexPatternChange.bind(this, controlIndex)}
```
These binds too should be moved to constructor or es7 syntax (and below)

Stacey-Gammon replied:
Question for @weltenwort and @kjbekkelund (and perhaps @cjcenizal when you get back) - 

This is actually a more interesting example because it's not just a simple bind but binds the value.  You could resolve this by passing `controlIndex` into `handleIndexPatternChange` inside the `ListControlEditor` , `RangeControlEditor`, and `ControlEditor ` components.  So something like:

```
export function ListControlEditor(props) {
  function handleIndexPatternChange() {
    props.handleIndexPatternChange(props.controlIndex);
  }
   return (
     <div>
       <IndexPatternSelect
         value={props.controlParams.indexPattern}
         onChange={handleIndexPatternChange}
         getIndexPatterns={props.getIndexPatterns}
      />
....
```

which resolves the performance issues of binding in render (Well, I think, I'm a little confused about how to do that with a functional component.  I'm not 100% sure my version above does not become the equivalent of binding in render if it was using class syntax).  But then you have all this extra code for every specialized type of component.

Have you guys come across this before in your React travels and have any words of wisdom, or preferences for how you would handle this?  Ignore the performance implications and just do the easy bind in render? Or write extra code to avoid it? Or something else I haven't thought of?

weltenwort replied:
Unfortunately your solution would have the same effect as using `.bind()` in that both create a new function on every render call. I can think of a few options:

* cache the handlers with `controlIndex` partially applied (I did this with [recompose's `withHandlers` HOC](https://github.com/acdlite/recompose/blob/master/docs/API.md#withhandlers) in other projects, but I think we don't yet have that incredibly useful dependency yet :wink: could be implemented in the stateful component via `ReceiveProps`)
* expect the `ListControlEditor` to supply its `controlIndex` when calling its `props.handleIndexPatternChange` (this couples the components a bit stronger)
* somehow store the index in the doc and retrieve it from the `event.target` (messy, would consider this the last options)
* just accept that it creates new functions (if the component is pure and rarely re-rendered I wouldn't expect this to cause problems)

Stacey-Gammon replied:
Nevermind @weltenwort and @kjbekkelund  (and @cjcenizal) - @nreese is a step ahead of me and came up with an elegant solution.

weltenwort replied:
nice, what does it look like?

Stacey-Gammon commented on src/core_plugins/input_control_vis/public/components/vis/list_control.js:
```diff
@@ -0,0 +1,48 @@
+          simpleValue={true}
+          value={this.props.control.value}
+          options={this.props.control.selectOptions}
+          onChange={this.handleOnChange.bind(null, this.props.control)}
```
move bind to constructor

Stacey-Gammon commented on src/core_plugins/input_control_vis/public/vis.less:
```diff
@@ -0,0 +1,22 @@
+  }
+}
+
+.Select-menu-outer {
```
This is how you are making the pop ups appear in front of the other panels, right?  Just a thought but this is also why putting a generic select that wraps react-select in the uiFramework would be nice.  These kinds of css customizations can be kept in there, and then anywhere a drop down select needs to be used, it can take advantage of the custom css living in the uiFramework, not in the visualize plugin.  I'll leave it up to you though, can always be something tackled later, post k7.

btw, the capitalization is bc of react-select, not us right?

JacobBrandt commented:
Nice work.  This is going to be very useful.

Two things to consider
1) With a large difference between min and max and very small steps it becomes increasingly difficult to use the slider to get to the value you want.
2) When you get to a scenario where you want the min and max to be within a few steps of one another over a range of thousands of steps the two dots end on top of each other providing a less then friendly UX.  You can't see both values and it's difficult to select the correct one to change.

Possible solution:
Create two sliders for both Min and Max.  Min slider value controls min value for Max slider.  Add input field to both so you can type in the values you want.

nreese commented:
@Stacey-Gammon @snide @kobelb What do you think about @JacobBrandt's idea of ditching `react-input-range` and just using 2 native range inputs? The latest commit changes the range control to provide a very rough implementation showing how @JacobBrandt's request would function. Is this better, worse, or just different? Is this worth pursuing further? I can polish this up if this seems like a better control mechanism.

<img width="631" alt="screen shot 2017-09-15 at 8 39 05 am" src="https://user-images.githubusercontent.com/373691/30487811-c5f36a3a-99f1-11e7-8ebc-64b951f680d7.png">

kobelb commented:
@nreese to be completely honest, I don't find it to be intuitive at all...

JacobBrandt commented:
@nreese Now that I see it I can see what @kobelb is saying.  Maybe try putting the two input fields side by side in the middle.

<-------\*---->  7105 to 8771 <---*-------->

Another option is to keep react-input-range and put the input fields at the beginning or end of the slider

<------\*-------------------*--> 7105 to 8711

You'll still run into issues where the dots are on top of each other but with the input controls you'll at least be able to see the values you can't read on the slider.

snide commented:
![image](https://user-images.githubusercontent.com/324519/30501607-f0999ea0-9a17-11e7-8242-eb689b837ee3.png)

JacobBrandt commented:
@snide Yea that's better if we're going to keep the one slider approach.

Stacey-Gammon commented:
I like the single slider better too. ++ to @snide's suggestion.

nreese commented:
@Stacey-Gammon Thanks for all the great feedback. Below are responses. I tried to summarize here instead of answering each individually. Would you mind taking another look at the PR?
 * **`Options` tab opened by default:** Great point, I have changed so that the `Controls` tab is now the default tab when the editor is opened.
 * **Adjust the indexPattern drop down from allowing any dynamic index pattern:** This is not possible. The control must target a single index pattern.
 * **Bug with the index pattern drop down where enter a character that doesn't match any index pattern, click outside of the control, and finally press the drop down icon to see the possible index patterns, and nothing shows up:** This is a bug with react-select and will need to be fixed upstream. I will create a PR for react-select and get this fixed.
* **Padding above the control buttons:** Padding has been added
* **The difference between Clear and Cancel isn't very clear:** Changed buttons to `Apply Changes`, `Cancel Changes`, and `Clear Form`. Modified how `Clear Form` works so that instead of clearing kibana filters, it just clears the form and `Apply Changes` still has to be clicked to clear the kibana filters.
* **delay when pressing Clear:** There is no more delay when `Clear Form` is clicked.
* **Unable to select certain options in list control drop down:** Good find - there was a bug where numeric values or string values containing a comma were un-selectable. That has been resolved.
* **input controls can also extend the borders of the dashboard panels**: Unresolved and tracked by issue https://github.com/elastic/kibana/issues/13982

nreese commented:
@JacobBrandt I have updated the range control to include inputs for setting the value more precisely. Try it out and see if this makes sense

Stacey-Gammon commented:
Awesome, the new buttons and labels are way more intuitive.  One little UX oddity, but does not need to be addressed in this PR:

![slider](https://user-images.githubusercontent.com/16563603/30560934-67571e34-9c87-11e7-9d36-23ff4a2d08f9.gif)

JacobBrandt commented:
@nreese Yep. This makes sense and will work well, thanks!

nreese commented:
@timroes Everything should be good to go. Can you take another look?

nreese commented:
backported to 6.x https://github.com/elastic/kibana/pull/14074

jccq commented:
my2c., making this work on top of a saved search instead (or additionally to ) an index pattern could also be useful (Future version?)

nreese commented:
@jccq interesting idea, added to feature request meta issue https://github.com/elastic/kibana/issues/13911
